packages/CodingStandard/src/Fixer/DependencyInjection/InjectToConstructorInjectionFixer.php

```diff
@@ -0,0 +1,239 @@
+            [
+                new CodeSample(
+'<?php
+class SomeClass
+{
+    /**
+     * @inject
+     * @var stdClass
+     */
+    public $property;
+}'),
+            ]
+        );
+    }
+
+    public function isCandidate(Tokens $tokens): bool
+    {
+        return $tokens->isAllTokenKindsFound([T_CLASS, T_DOC_COMMENT, T_PUBLIC, T_VARIABLE]);
+    }
+
+    public function fix(SplFileInfo $file, Tokens $tokens): void
+    {
+        // 1. find annotation @inject
+        foreach ($tokens as $index => $token) {
+            if (! $token->isGivenKind(T_DOC_COMMENT)) {
+                continue;
+            }
+
+            $doc = new DocBlock($token->getContent());
+            $injectAnnotations = $doc->getAnnotationsOfType('inject');
+            if (! count($injectAnnotations)) {
+                continue;
+            }
+
+            // 1. remove it
+            foreach ($injectAnnotations as $injectAnnotation) {
+                $injectAnnotation->remove();
+            }
+
+            $token->setContent($doc->getContent());
+
+            // 2. make public property private
+            for ($i = $index;; ++$i) {
+                $token = $tokens[$i];
+                if ($token->isGivenKind(T_PUBLIC)) {
+                    $token->override([T_PRIVATE, 'private']);
+                    break;
+                }
+            }
+
+            // 3. add dependency to constructor
+            $propertyName = '';
+            for ($i = $index;; ++$i) {
+                $token = $tokens[$i];
+                if ($token->isGivenKind(T_VARIABLE)) {
+                    $propertyName = ltrim($token->getContent(), '$');
+                    break;
+                }
+            }
+            $varAnnotation = $doc->getAnnotationsOfType('var')[0];
+            $propertyType = $varAnnotation->getTypes()[0];
+
+            // detect constructor
+            $constructMethodPosition = null;
+            for ($i = $index; $i < count($tokens); ++$i) {
+                $token = $tokens[$i];
+                if ($token->isGivenKind(T_FUNCTION)) {
+                    $namePosition = $tokens->getNextNonWhitespace($i);
+                    $methodNameToken = $tokens[$namePosition];
+                    if ($methodNameToken->getContent() === '__construct') {
+                        $constructMethodPosition = $i;
+                        break;
+                    }
+                }
+            }
+
+            // A. has a constructor?
+            if (is_int($constructMethodPosition)) { // "function" token
+                $this->addPropertyToConstructor($tokens, $propertyType, $propertyName, $constructMethodPosition);
+            } else {
+                // B. doesn't have a constructor?
+                $this->addConstructorMethod($tokens, $propertyType, $propertyName);
+            }
+
+            // run again with new tokens; @todo: can this be done any better to notice new __construct method added
+            // by these tokens?
```

Comment:
Thanks for your compassion. I'm slowly getting used to this. It looks there is quite low code coverage for the core code of PHP-CS-Fixer.

It's really painful to find "use x instead" myself by error and trial instead of in code of that package. Deprecation messages like "This is deprecated" have no value at all.

I try to put these into debug messages with link, so it's easy to work with them. Like

```php
/**
 * @deprecated This is deprecated since x.y, will be removed.
 * Use @see SomeClass::someMethod() instead.
 */
```